### PR TITLE
Add Redux

### PR DESCRIPTION
**Pull request by jhb188**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,3 @@
+NODE_ENV=development
+PORT=3000
+NODE_PATH=src/
```
Comment (jhb188) [inline .env.example]: this lets us import from `components/SongList`, etc.

File: src/components/NewSongForm.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ class NewSongForm extends Component {
   saveSong (e) {
     e.preventDefault()
     const {title, artist, album, votes} = this.state
-    Firebase.database().ref('songs').push({title, artist, album, votes})
+    this.props.firebase.push('/songs', {title, artist, album, votes})
```
Comment (jhb188) [inline src/components/NewSongForm.js]: would like to hide the implementation details of `saveSong` from this component and make that function available as a prop as was done for the `changeVote` function

File: src/components/SongList/SongList.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import React, { Component } from 'react'
```
Comment (jhb188) [inline src/components/SongList/SongList.js]: separated some of the state logic `SongList/index.js` from rendering logic `SongList/SongList.js`. ideally, we can pull the remaining state in `SongList.js` into another redux slice/reducer.

File: src/components/SongList/index.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { compose } from 'redux'
+import { connect } from 'react-redux'
+import { firebaseConnect } from 'react-redux-firebase'
+import SongList from './SongList';
+
+const sortDescByVotes = compose(
+  reverse,
+  sortBy(prop('votes')),
+  Object.values,
+  mapObjIndexed((song, id) => ({ ...song, id }))
```
Comment (rossvz) [inline src/components/SongList/index.js]: this is fantastic... I really need to dive deeper into Ramda, this is such a nice succinct way of writing this.

File: src/components/SongList/index.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  connect(
+    ({ firebase }) => ({
+      songs: sortDescByVotes(firebase.data.songs),
+    }),
+    {},
+    (stateProps, dispatchProps, ownProps) => ({
+      ...stateProps,
+      ...dispatchProps,
+      ...ownProps,
+      changeVote: song => value => {
```
Comment (rossvz) [inline src/components/SongList/index.js]: ideally, would this function live somewhere else? As we add more actions I could see this getting messy

Comment (jhb188) [reply src/components/SongList/index.js]: Yes, good point. I’m going to write this as a redux action creator in a subsequent PR.

**Pull request merged**